Add Mega: Moving Average Equipped Gated Attention

mnaylor5 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #19982 

This pull request adds [Mega: Moving Average Equipped Gated Attention](https://arxiv.org/abs/2209.10655), which is the current leader of the [LRA benchmark](https://paperswithcode.com/sota/long-range-modeling-on-lra). Adapted from the original [fairseq-based repo](https://github.com/facebookresearch/mega) and used a MLM checkpoint I created using the original implementation on the wikitext-103 dataset. There is no proposed Mega tokenizer, so I used the RoBERTa tokenizer which I used on the wikitext checkpoint. The proposed implementation works in encoder and decoder settings, and all relevant tests are passing.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker and @younesbelkada for text models; tagging @NielsRogge for visibility as he responded to the original issue.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

mnaylor5 committed:
add mega file structure and plain pytorch version of mega source code

mnaylor5 committed:
added config class with old naming conventions

mnaylor5 committed:
filled in mega documentation

mnaylor5 committed:
added config class and embeddings with optional token types

mnaylor5 committed:
updated notes

mnaylor5 committed:
starting the conversion process, deleted intermediate and added use_cache back to config

mnaylor5 committed:
renamed config attributes in modeling_mega.py

mnaylor5 committed:
checkpointing before refactoring incremental decoding functions

mnaylor5 committed:
removed stateful incremental key/values for EMA and self-attention

mnaylor5 committed:
refactored MovingAverageGatedAttention to remove stateful k/v history and use unified attention mask

mnaylor5 committed:
MovingAverageGatedAttention works with incremental decoding + past values, added sequence length enforcement

mnaylor5 committed:
more comments in MovingAverageGatedAttention + checkpointing before GatedCrossAttention

mnaylor5 committed:
Merge remote-tracking branch 'upstream/main' into add-mega

mnaylor5 committed:
bug fix in attention mask handling in MovingAverageGatedAttention

mnaylor5 committed:
removed incremental state from GatedCrossAttention and removed IncrementalState class

mnaylor5 committed:
finished gated cross attention and got MegaLayer working

mnaylor5 committed:
fixed causal masking in mega decoder

mnaylor5 committed:
fixed how padding and causal masks are passed through MegaLayer with and without k/v caching

mnaylor5 committed:
finished MegaModel; tested with encoder, decoder-only, and cross-attention type inputs; started work on downstream classes; removed mentions of position_ids

mnaylor5 committed:
added optional dense hidden layer for masked and causal LM classes

mnaylor5 committed:
docstring updates in MultiHeadEMA and GatedCrossAttention, removed unnecessary inputs in cross-attention

mnaylor5 committed:
removed before_attn_fn in Mega class and updated docstrings and comments up to there

mnaylor5 committed:
bug fix in MovingAverageGatedAttention masking

mnaylor5 committed:
working conversion of MLM checkpoint in scratchpad script -- perfect matches

mnaylor5 committed:
moved arg for hidden dense layer in LM head to config; discovered issue where from_pretrained is renaming gamma and beta parameters

mnaylor5 committed:
renamed gamma and beta parameters to avoid HF renaming when loading from checkpoint

mnaylor5 committed:
Merge remote-tracking branch 'upstream/main' into add-mega

mnaylor5 committed:
finished checkpoint conversion script

mnaylor5 committed:
cleanup old class in mega config script

mnaylor5 committed:
removed 'copied from' statements and passing integration tests

mnaylor5 committed:
added num_attention_heads=1 to config for integration compatibility, decoder tests working, generation tests failing

mnaylor5 committed:
fixed tuple output of megamodel

mnaylor5 committed:
all common tests passing after fixing issues in decoder, gradient retention, and initialization

mnaylor5 committed:
added mega-specific tests, ready for more documentation and style checks

mnaylor5 committed:
updated docstrings; checkpoint before style fixes

mnaylor5 committed:
Merge remote-tracking branch 'upstream/main' into add-mega

mnaylor5 committed:
style and quality checks, fixed initialization problem in float_tensor, ready for PR

mnaylor5 committed:
added mega to toctree

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21766). All of your documentation changes will be reflected on that endpoint.

mnaylor5 committed:
removed unnecessary arg in megaconfig

mnaylor5 committed:
removed unused arg and fixed code samples with leftover roberta models

mnaylor5 commented:
Sorry for the initial test failures! It should be taken care of now. Also I wanted to point out that I did not have access to a GPU while developing this, so I was not able to test on a GPU

younesbelkada commented:
Hi @mnaylor5 
Thanks for your great work on this! Let us know when do you think this is ready for review 💪

mnaylor5 commented:
Thank you @younesbelkada! It is ready for review now 😄

mnaylor5 commented:
Hi @younesbelkada / @ArthurZucker - just checking in to see if there is anything you need from me before reviewing this pull request. Looking forward to being able to use Mega in `transformers`!

ArthurZucker commented:
Hey! I'll give you a review tomorrow! Sorry for the wait, had to synch with @younesbelkada on this one

mnaylor5 commented:
Thanks @ArthurZucker, and no worries! 😄

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        nn.init.normal_(self.fc2.weight, mean=0.0, std=std)
+        nn.init.constant_(self.fc2.bias, 0.0)
+
+    def forward(self, x):
+        residual = x
+
+        if self.prenorm:
+            x = self.norm(x)
+
+        x = self.activation(self.fc1(x))
+        x = self.hidden_dropout(x)
+        x = self.fc2(x)
+        x = self.dropout(x)
+        x = x + residual
+
+        if not self.prenorm:
+            x = self.norm(x)
+
+        return x
```
better naming for these variables

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        self.fc1 = nn.Linear(self.config.hidden_size, self.config.nffn_hidden_size)
+        self.fc2 = nn.Linear(self.config.nffn_hidden_size, self.config.hidden_size)
+
+        self.reset_parameters()
```
unless this is specific to MEGA as you would need to run this during inference and training, should not be needed

mnaylor5 replied:
noted!

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        attn_weights = self.softmax(qk).type_as(qk)
+        return attn_weights
+
+    def forward(
```
no single letter variables

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        # C x C
+        bias = self.rel_pos_bias(slen)
+        if slen != q.size(2):
+            assert q.size(2) == 1, "Size mismatch between Q and K in softmax attention"
```
no assedrt, let's raise an error

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        )
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        # possibly remove this in favor of automated initialization by HF classes
+        std = self.config.initializer_range
+        mean = 0.0
+        nn.init.normal_(self.v_proj.weight, mean=mean, std=std)
+        nn.init.constant_(self.v_proj.bias, 0.0)
+
+        nn.init.normal_(self.mx_proj.weight, mean=mean, std=std)
+        nn.init.constant_(self.mx_proj.bias, 0.0)
+
+        nn.init.normal_(self.h_proj.weight, mean=mean, std=std)
+        nn.init.constant_(self.h_proj.bias, 0.0)
+
+        nn.init.normal_(self.g_param, mean=mean, std=std)
+        nn.init.constant_(self.b_param, 0.0)
```
yes let's remove 😉

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        self.module_name = module_name
+        self.apply_during_inference = False
+
+    def forward(self, x, batch_first: bool = False, inplace: bool = False):
```
no assserts and better variable names please!

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        return x
+
+
+class SequenceNorm(nn.Module):
```
SHould be `MegaSequenceNorm`

ArthurZucker replied:
Also should be added to the list of normalisation layers that we have `ALL_LAYERNORM_LAYERS`

mnaylor5 replied:
Sounds good! I'll rename and add it there

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        if self.affine:
+            nn.init.constant_(self.scalar, 1.0)
+
+    def forward(self, x):
+        mean_square = torch.mean(torch.square(x), dim=self.dim, keepdim=True)
+        if self.scalar is not None:
+            x = self.scalar * x
+
+        x = x * torch.rsqrt(mean_square + self.eps)
+        return x
+
+
+class RMSNorm(nn.Module):
+    def __init__(self, number_features, eps=1e-6, affine=True):
+        super().__init__()
+        self.num_features = number_features
+        self.eps = eps
+        self.affine = affine
+        if affine:
+            self.weight = nn.Parameter(torch.Tensor(self.num_features))
+        else:
+            self.register_parameter("weight", None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        if self.affine:
+            nn.init.constant_(self.weight, 1.0)
+
+    def forward(self, x):
+        mean_square = torch.mean(torch.square(x), dim=-1, keepdim=True)
+        if self.weight is not None:
+            x = x * self.weight
+
+        x = x * torch.rsqrt(mean_square + self.eps)
+        return x
```
are there required?

mnaylor5 replied:
yes, these are selected based on the option selected in [normalization_type](https://github.com/mnaylor5/transformers/blob/add-mega/src/transformers/models/mega/configuration_mega.py#L71-L73)

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        return t
+
+
+class RotaryRelativePositionalBias(nn.Module):
```
Quick question, what is this based on? (we have already 3 implementations of Rotary position embedding, would be nice to have a reference (I mean is this from their repo etc)

ArthurZucker replied:
One example [here](https://github.com/ArthurZucker/transformers/blob/fix-zac-pipleine/src/transformers/models/esm/modeling_esm.py#L87)

mnaylor5 replied:
Apologies, that could have been clearer. This is based on the original implementation! See https://github.com/facebookresearch/mega/blob/09f1a872c77bdf78e1b128c21c002f08fd6cc2dc/fairseq/modules/relative_positional_bias.py#L49

ArthurZucker replied:
Edit, after reading the paper, it seems that the `relative positional
bias` is based on the `T5` , if we can have copied from would be great! Otherwise let's just add MegaRotaryRelativePositionalBias

ArthurZucker replied:
Also let's move them to the top of the file 😉

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+
+
+# Positional embeddings
+# copied without modification from original Mega code
```
no single letter variable and no `reset`.

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+            key_padding_mask = None
+
+        if key_padding_mask is not None:
+            assert key_padding_mask.size(0) == bsz
+            assert key_padding_mask.size(1) == ctx_len
```
No assert in the codebase

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        """
+        EMA-based self-attention
+
+        Inputs
+            query (Tensor): the self (or target) sequence input with shape (Target Sequence Length x Batch x Embedding
+            Size) key (Tensor): the cross (or source) sequence input with shape (Source Sequence Length x Batch x
+            Embedding Size) value (Tensor): the cross (or source) sequence input with shape (Source Sequence Length x
+            Batch x Embedding Size) padding_mask (Tensor): NOT USED, but left in place to match the original
+            implementation key_padding_mask (optional, Tensor): padding mask corresponding to the source sequence
+            (Batch Size x Source Sequence Length) prev_key_values (tuple of Tensors, optional): if provided, the hidden
+            state returned from the previous
+                timestep during incremental decoding; expects that prior cross-attention keys and values will be the
+                last two items in the tuple
+            output_attentions (boolean): if true, cross-attention weights will be returned use_cache (boolean): if
+            True, perfom incremental decoding; uses `prev_state` as the prior
+                timestep, and returns the updated EMA hidden state for use in the next step
```
Let's  follow the correct docstring format here

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+        """
+
+        seq_len, bsz, embed_dim = query.size()
+        assert embed_dim == self.config.hidden_size
```
No new assert!

ArthurZucker reviewed (commented):
Thanks a lot for contributing! This models looks very interesting 😉 
As a first big modification, we should remove the `rest_param` in favor of HF `_init_weights`. Also let's find proper variable names for `x` to make the code more readable!

mnaylor5 commented:
Thanks for the review @ArthurZucker! I'll reply to individual comments where I can clear things up, and I'll accept your suggestions wherever I can. I'll probably be able to start on the modifications later today, and if not, then early next week.

mnaylor5 commented on src/transformers/models/mega/configuration_mega.py:
```diff
@@ -0,0 +1,239 @@
+# coding=utf-8
+# Copyright 2023 The Google AI Language Team Authors and The HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+#
```
Ah, good catch 🤦‍♂️

mnaylor5 commented on src/transformers/models/mega/configuration_mega.py:
```diff
@@ -0,0 +1,239 @@
+    Args:
+        vocab_size (`int`, *optional*, defaults to 30522):
+            Vocabulary size of the Mega model. Defines the number of different tokens that can be represented by the
+            `inputs_ids` passed when calling [`MegaModel`] or [`TFMegaModel`].
```
You are correct - that was left over from the RoBERTa model-like command

mnaylor5 commented on src/transformers/models/mega/convert_mega_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,268 @@
+        # which is the opposite of what HF returns
+        if ignore_mask_value == 0:
+            attention_mask = 1 - attention_mask
+
+        # get token embeddings from IDs
+        embeds = self.embedding_layer(input_ids)
+
+        # pass through the Mega layers
+        # input is (time, batch, encoder dim) and output is the same
+        for encoder in self.encoders:
+            embeds = encoder(embeds, attention_mask)
+
+        # return according to the shape specified
+        if batch_first:
+            # (T, B, H) --> (B, T, H)
+            return torch.transpose(embeds, 0, 1)
+        else:
+            return embeds
+
+
+# renamed from MegaForMaskedLM to avoid confusion with new module
+class OriginalMegaForMaskedLM(nn.Module):
+    "A wrapper class for doing masked language modeling with Mega"
+
+    def __init__(self, mega_args, depth, vocab_size):
+        super().__init__()
+        self.mega = MegaLM(mega_args, depth, vocab_size)
+        self.mlm_head = nn.Linear(mega_args.encoder_embed_dim, vocab_size)
+        self.dropout = nn.Dropout(p=0.1)
+
+    def forward(self, input_ids, attention_mask, batch_first=True, ignore_mask_value=0):
+        """
+        Perform a forward pass through the Mega encoder and the masked LM head. Returns logits for each vocabulary
+        entry.
+
+        If `batch_first` (default to align with Hugging Face tokenizer behavior), output will have the shape (Batch
+        size, Sequence length, Vocab size); otherwise (S, B, V)
+        """
+        encoder_output = self.mega(input_ids, attention_mask, batch_first, ignore_mask_value)
+        return self.mlm_head(self.dropout(encoder_output))
```
This was a really tricky part of the implementation process -- the checkpoint I converted was a BERT-style MLM that I created using the original repo and these two classes. This came up from a discussion in https://github.com/huggingface/transformers/issues/19982, because the original repo did not contain a BERT-like checkpoint; I trained the model in [this Colab notebook](https://colab.research.google.com/drive/1qfUO6o5HRdxBblWlw058HVyvaEPhPpH8?usp=sharing) and reused the code at the bottom in this checkpoint conversion script. 

I looked into using the exact model files the authors provide in the repo (which at the time was mostly LRA tasks and, imo, less generally helpful than something like a BERT model), but there wasn't a consistent way to instantiate the weights or tokenizers, and the authors haven't been responding to GitHub issues for a few months.

mnaylor5 commented on src/transformers/models/mega/convert_mega_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,268 @@
+    # load the original encoder
+    original_mlm = OriginalMegaForMaskedLM(**mega_original_args).eval()
+
+    # load its weights
+    print(
+        "Original Mega encoder:",
+        original_mlm.mega.load_state_dict(
+            torch.load(os.path.join(pretrained_checkpoint_path, "encoder_weights.pt"), map_location="cpu")
+        ),
+    )
+    print(
+        "Original Mega MLM layer:",
+        original_mlm.mlm_head.load_state_dict(
+            torch.load(os.path.join(pretrained_checkpoint_path, "mlm_head_weights.pt"), map_location="cpu")
+        ),
+    )
```
Ah, sorry - I was following the RoBERTa conversion script which loads the original model and copies the weights into the newly initialized one:

https://github.com/huggingface/transformers/blob/main/src/transformers/models/roberta/convert_roberta_original_pytorch_checkpoint_to_pytorch.py

mnaylor5 commented on src/transformers/models/mega/convert_mega_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,268 @@
+        ),
+    )
+
+    # test on a randomly generated input sequence
+    input_ids = torch.randint(0, hf_config.vocab_size, size=(4, 256))
+    input_mask = torch.ones_like(input_ids)
+    # mask a few tokens to make sure masking is applied appropriately :)
+    input_mask[:, -10:] = 0
+
+    # run forward passes
+    original_output = original_mlm(input_ids, input_mask, batch_first=True, ignore_mask_value=0)
+    hf_output = hf_mlm(input_ids, input_mask)[0]
+
+    # print shapes and diff
+    print(f"original output {original_output.shape}")
+    print(f"hf output {hf_output.shape}")
+    print(f"max diff: {(original_output - hf_output).max()}")  # 0.0
+    success = torch.allclose(original_output, hf_output, atol=1e-3)
```
Thanks! That's helpful context. This is another component that I followed from the RoBERTa script -- since that script tests the difference in outputs between the original and converted models, I assumed that was an expected part of the process:

https://github.com/huggingface/transformers/blob/main/src/transformers/models/roberta/convert_roberta_original_pytorch_checkpoint_to_pytorch.py#L142-L156

I think I also conflated steps 6 and 7 in the [contribution guidelines](https://huggingface.co/docs/transformers/add_new_model#514-port-brandnewbert-to-transformers), so when I saw that RoBERTa had the closeness assertion in its conversion script, I just followed suit.

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+
+
+# starting with activation functions
+def relu2(x):
+    return torch.square(F.relu(x))
```
That's fair :) 
This was something I copied directly from the original source, so if there is a preference to move away from this style, I can implement it as I go back through to make the rest of the style changes

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2299 @@
+    return 0.5 * x * (1 + torch.tanh(gelu_accurate._a * (x + 0.044715 * torch.pow(x, 3))))
+
+
+def get_activation_fn(activation: str) -> Callable:
+    """Returns the activation function corresponding to `activation`"""
+    if activation == "relu":
+        return F.relu
+    elif activation == "gelu":
+        return F.gelu
+    elif activation == "gelu_accurate":
+        return gelu_accurate
+    elif activation == "silu":
+        return F.silu
+    elif activation == "linear":
+        return lambda x: x
+    else:
+        raise RuntimeError("--activation-fn {} not supported".format(activation))
```
:+1: thanks!

mnaylor5 commented on tests/models/mega/test_modeling_mega.py:
```diff
@@ -0,0 +1,592 @@
+
+
+@require_torch
+class MegaModelIntegrationTest(TestCasePlus):
```
Okay! Are there some examples of the tests I should add? I followed [RoBERTa's integration tests](https://github.com/huggingface/transformers/blob/main/tests/models/roberta/test_modeling_roberta.py#L527) and removed the sequence classification test because there is currently no pretrained sequence classifier for Mega

mnaylor5 replied:
Also on this: there are some generation tests in the common modeling tests, so I assumed that since RoBERTa doesn't add any further generation tests, this would be okay for Mega

ArthurZucker replied:
Roberta is an old model, we try to have as many test as possible especially for new models! Take inspiration from T5 or OPt!

mnaylor5 replied:
Okay I'll take a look and add what I can!

mnaylor5 replied:
It looks like OPT is the closer of the two models, since it can behave as a standalone encoder or decoder like MEGA/BERT/etc.

I added the `test_generate_fp16` test from OPT, and it's passing locally on CPU. The other generation tests for OPT are testing the correctness of generated values from existing pretrained LMs (for example, [this one](https://github.com/huggingface/transformers/blob/main/tests/models/opt/test_modeling_opt.py#L435)). I held off on porting these over because there is no pretrained causal LM for Mega as of now.

ArthurZucker replied:
Thanks for adding that, looks good to me

mnaylor5 committed:
Apply suggestions from code review

Applied all suggestions except the one renaming a class, as I'll need to update that througout

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

mnaylor5 committed:
fixed issue where .view breaks batch dimension, conversion script fixed with absolute imports, updated readme with Mega->MEGA

mnaylor5 committed:
removed asserts in Mega code, renamed sequencenorm, gatedcrossattention, and NFFN, replaced get_activation_fn with ACTFN, and added sequencenorm to layer norms

mnaylor5 committed:
reformatted .forward() docstrings to match style and removed unused mask input in cross-attention

mnaylor5 committed:
removed all reset_parameters() methods and rolled into MegaPreTrainedModel._init_weights()

mnaylor5 committed:
renamed all single-letter variables and improved readability in tensor size comments, Mega->MEGA in 2 documentation files

mnaylor5 committed:
variable names in NFFN

mnaylor5 committed:
manual Mega->MEGA changes in docs

mnaylor5 committed:
Mega->MEGA in config auto

mnaylor5 committed:
style and quality fixes

mnaylor5 committed:
Merge remote-tracking branch 'upstream/main' into add-mega

mnaylor5 commented:
Alright @ArthurZucker this should be good to review again! The biggest updates in this version are removing the `reset_parameters` methods in favor of `_init_weights`, renaming variables/comments to avoid single-letter names, docstring format updates, and renaming `Mega` to `MEGA` based on your suggestion. I have resolved the comments where I made the changes, and left the other comments in place for continued discussion.

Thanks again for your feedback, and I'm happy to answer any questions that arise. Looking forward to getting MEGA into the library! 🚀

mnaylor5 commented:
Hi there @ArthurZucker - thanks again for the feedback in your previous review. Just reaching out to see if anything else is needed before reviewing and hopefully merging!

ArthurZucker commented:
Hey! Sorry I must have missed your previous ping! Will review now!

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+            past_ema_state = None
+            past_vandermonde = None
+
+        # (kernel_dim X ema_projection_size X sequence_length)
+        vander = vander[:, :, :-1]
+        kernel = (p_coeff * self.b_param) * vander
+        kernel_proj = torch.einsum("dnl,dn->dl", kernel, self.g_param * self.scale)
+
+        kernel_fourier = torch.fft.rfft(kernel_proj.float(), n=2 * length)
+        inputs_fourier = torch.fft.rfft(inputs.float(), n=2 * length)
+        # (batch_size X kernel_dim X sequence_length)
+        out = torch.fft.irfft(inputs_fourier * kernel_fourier, n=2 * length)[..., 0:length]
```
This is probably the `fast_convolution` algorithm, let's create a function to add some doc on where this comes from 😉

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        # (batch_size, kernel_dim, ema_projection_size)
+        return out.permute(2, 0, 1), updated_hidden_state
+
+    def one_step(self, inputs, past_state=None):
```
Is this `one_ema_step`

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        # (1 X batch_size X kernel_dim), (batch_size X kernel_dim X ema_projection_size)
+        return out.unsqueeze(0), updated_state
+
+    def forward(
```
Let's split this into sub layers. We can have `MegaDampedEMALayer`, then the `MegaGatedAttentionUnit` and `MegaSingleHeadAttentionLayer`. This will make the code more understanble w.r.t the paper IMO. Then the `MegaLayer` is the final wrapper.

ArthurZucker replied:
Also there is no self attention in this layer (the doc says so)

mnaylor5 replied:
Docstring updated in upcoming commit - after looking back at this, I think this `MultiDimensionDampedEMA` class is fairly self-contained as it's only performing the EMA component, and it aligns with the paper's EMA sub-layer.

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        return output
+
+
+class RMSNorm(nn.Module):
```
This is the same as T5LayerNorm no? let's add copied from (T5 norm is a RMSNorm AFAIK)

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        return input
+
+
+class MegaSequenceNorm(nn.Module):
```
Small nit, note really a big fan of this, would rather just index over `NORM2FN` especially because the paper only uses 1. (people could choose the layer norm and we just have to do something along the lines of 
```python 
self.layer_norm = NORM2FN[config.norm]` 
```
like we do for the activation fuctions 😉 see [here](https://github.com/ArthurZucker/transformers/blob/ArthurZucker/issue21300/src/transformers/models/mobilebert/modeling_mobilebert.py#L170)

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        self.p = p
+
+    def forward(self, input, batch_first: bool = False, inplace: bool = False):
+        if self.training:
+            return F.dropout(input, p=self.p, training=True, inplace=inplace)
+        else:
+            return input
```
when the model is not training, dropout is not applied no? We can probably remove this entirely

mnaylor5 replied:
Done and coming in next commit

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+            return input
+
+
+class MegaFeatureDropout(nn.Module):
```
Same comment here

mnaylor5 replied:
Same here

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        return output
+
+
+class MegaEmbeddings(nn.Module):
```
Roberta is a pretty old model! Usually this is done in the `MegaModel` now as the checks for input ids and embeds should be decoupled from here

ArthurZucker replied:
Also, token_type_ids should be declared as a buffer : 
```python 
        # Setting the token_type_ids to the registered buffer in constructor where it is all zeros, which usually occurs
        # when its auto-generated, registered buffer helps users when tracing the model without passing token_type_ids, solves
        # issue #5664
```

mnaylor5 replied:
Added for upcoming commit!

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+
+        return outs
+
+
```
Missing copied from here

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        self.config = config
+
+        self.embedding_layer = MegaEmbeddings(config)
+        self.encoders = nn.ModuleList([MegaLayer(config) for _ in range(config.num_hidden_layers)])
```
We can also wrap this in a MegaEncoder,. (Bert is a good inspiration on that).
This is in order to run the common tests!

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        )
+
+
+class MegaClassificationHead(nn.Module):
```
unless this is copied from, no single letter variables

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
I did not really check the docstring will do that in a final review 😉

ArthurZucker reviewed (commented):
Hey! Awesome work! The model is very interesting, left a few comments, mostly for compatibility with the transformer philosophy, but really good job on implementing this! 
Ping me whever for another review, this is taking a good shape!

ArthurZucker reviewed (commented):
Hey! Awesome work! The model is very interesting, left a few comments, mostly for compatibility with the transformer philosophy, but really good job on implementing this! 
Ping me whever for another review, this is taking a good shape!

mnaylor5 commented on src/transformers/models/mega/configuration_mega.py:
```diff
@@ -0,0 +1,239 @@
+            Dimensionality of the linear projection for shared representation after silu gating (Z in the Mega paper).
+        use_chunking (`bool`, *optional*, defaults to `False`):
+            Whether to chunk inputs for linear self-attention complexity (described as Mega-chunk in the paper)
+        chunk_size (`int`, *optional*, defaults to -1):
+            If `use_chunking` is True, determines the size of the chunks to apply to the input sequence. If chunking is
+            used, input sequences must be padded to a multiple of `chunk_size`
+        truncation (`int`, *optional*, defaults to None):
+            If specified, the sequence length for which to truncate MultiHeadEMA
```
No problem!

mnaylor5 committed:
Apply suggestions from code review

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

mnaylor5 commented:
Thanks @ArthurZucker! I appreciate the quick review and the encouragement 😄 
I added a couple of questions where things weren't totally clear to me, but I can get started on everything else now. I'm really excited about getting this model into the library, and hopefully there won't be too many more changes required!

ArthurZucker commented:
Will answer to your questions tomorrow!

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        return embeddings
+
+
+class MegaLayer(nn.Module):
+    def __init__(self, config: MegaConfig):
+        super().__init__()
+        self.seq_len_dim = 1
+        self.mega_layer = MovingAverageGatedAttention(config)
```
thanks! renaming `MegaLayer->MegaBlock` to align with the paper's usage of "MEGA block" vs. "MEGA layer" throughout Section 3

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        self.delta = nn.Parameter(torch.Tensor(kernel_dim, self.ndim, 1))
+        self.alpha = nn.Parameter(torch.Tensor(kernel_dim, self.ndim, 1))
+        # renamed gamma and beta to g_param and b_param respectively to avoid HF renaming things
+        self.b_param = nn.Parameter(torch.Tensor(kernel_dim, self.ndim, 1))
```
Renamed - see other thread :+1:

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+    # See all Mega models at https://huggingface.co/models?filter=mega
+]
+
+# Mega source code converted to pure PyTorch
+# resources
+#   - paper: https://arxiv.org/abs/2209.10655
+#   - original implementation: https://github.com/facebookresearch/mega
+# notable differences from the original implementation:
+#   - refactored away from stateful representation of incremental decoding
+#     state in favor of Hugging Face's typical `past_key_values`
+#   - fixed inconsistency in how causal masks are expected by `softmax_attention`
+#     and `element_attention` in MovingAverageGatedAttention (see https://github.com/facebookresearch/mega/issues/11)
+#   - added support for token type embeddings (not specifically included
+#     or excluded in the original implementation/paper)
```
moved in upcoming commit!

mnaylor5 committed:
renamed parameters and variables with confusing names, added copied from statements, moved fft conv to its own method, other cleanup from PR comments

mnaylor5 committed:
commit before dealing with merge conflicts

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+
+# starting with activation functions
+# squared-relu and laplace are alternatives to softmax for attention activation
+def relu2(x):
+    relu_x = F.relu(x)
+    squared = torch.square(relu_x)
+    return squared
+
+
+def laplace(x, mu=0.707107, sigma=0.282095):
+    x = (x - mu).div(sigma * math.sqrt(2.0))
+    return 0.5 * (1.0 + torch.erf(x))
+
+
+# gelu-accurate is an alternative to gelu and is used with the remaining hidden activation functions in the
+# original MEGA repo, which all have exact equivalents in ACT2FN
+def gelu_accurate(x):
+    if not hasattr(gelu_accurate, "_a"):
+        gelu_accurate._a = math.sqrt(2 / math.pi)
+    return 0.5 * x * (1 + torch.tanh(gelu_accurate._a * (x + 0.044715 * torch.pow(x, 3))))
+
+
```
planning to address this after dealing with the merge conflicts that have crept in

mnaylor5 replied:
Done and included in latest commit 🙌 
also included new tests explicitly for each of these attention activation functions

mnaylor5 committed:
resolving merge conflicts

mnaylor5 committed:
made new attention activation functions available in ACT2FN and added generation test from OPT

mnaylor5 committed:
style and quality in activations and tests

mnaylor5 commented:
Alright @ArthurZucker, I think that's everything except the threads with ongoing discussion. I'm super happy with how this is shaping up! In the latest batch of commits:
* Renamed classes, variables, and params based on comments (mainly in EMA and MovingAverageGatedAttention class)
* Rearranged positional bias, normalization functions, activation functions, dropout classes
* Added the `copied from comments` where requested
* Added token type ID buffer
* Added tests for generation and sequence classification
* Moved FFT convolution into a reusable method with additional documentation
* Addressed merge conflicts from LLaMA 🦙 

Thanks for the feedback and I'll wait on any more changes until you get a chance to review the updates and resolve the open discussions. Excited to get up and running with MEGA in `transformers` 🚀 🤗

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2280 @@
+        return return_values
+
+
+# Normalized feed-forward network
+# left as-is from original Mega repo aside from retrieving args from Hugging Face config
+class MegaNormalizedFeedForwardNetwork(nn.Module):
```
Same comment here, let's move the comment to the docstring

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2280 @@
+        feed-forward layer
+
+        Args:
+            hidden_states (`torch.Tensor`` of shape `(target_sequence_length, batch_size, hidden_size)`):
```
```suggestion
            hidden_states (`torch.Tensor` of shape `(target_sequence_length, batch_size, hidden_size)`):
```

mnaylor5 replied:
Same here - fixing locally

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2280 @@
+    Equipped Gated Attention*_ by Xuezhe Ma, Chunting Zhou, Xiang Kong, Junxian He, Liangke Gui, Graham Neubig,
+    Jonathan May, and Luke Zettlemoyer
+
+    To behave as an decoder the model needs to be initialized with the `is_decoder` argument of the configuration set
```
```suggestion
    To behave as a decoder the model needs to be initialized with the `is_decoder` argument of the configuration set
```

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2280 @@
+        self.config = config
+
+        self.embedding_layer = MegaEmbeddings(config)
+        self.encoders = nn.ModuleList([MegaBlock(config) for _ in range(config.num_hidden_layers)])
```
Maybe `self.layers` is better suited 😉

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2280 @@
+        self.softmax = nn.Softmax(dim=-1)
+
+    def element_attention(self, query, key, key_padding_mask, pidx):
+        bsz, clen, _ = key.size()
+        slen = query.size(1) if pidx is None else pidx + 1
```
We usually use `tgt_len` and `src_len ` for `clen` and `slen`, let's stick to that!

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2280 @@
+        return tile
+
+
+# rotary positional bias similar to RoPE
+# taken from original Mega repo
+class RotaryRelativePositionalBias(nn.Module):
```
Also we recently had a problem with `padding = "left"` when using the rotary embedding ( see #22161 )
If this is similar to `RoPE`, we have it in transformers so using `# Copied from` would also work well and prevent from adding new code, but when I checked, it seems that we just compute a bias and add it to the query and key states, vs scale the query and key

mnaylor5 replied:
That's super interesting! It looks like the Mega implementation is susceptible to the same problem:

```python
from transformers import AutoTokenizer, MegaConfig, MegaModel
import torch
tok_left = AutoTokenizer.from_pretrained("mnaylor/mega-base-wikitext", padding_side="left")
inputs_left = tok_left(["This is an example tokenized input with padding"], return_tensors="pt", padding="max_length", max_length=25)

tok_right = AutoTokenizer.from_pretrained("mnaylor/mega-base-wikitext")
inputs_right = tok_right(["This is an example tokenized input with padding"], return_tensors="pt", padding="max_length", max_length=25)

model = MegaModel(MegaConfig(vocab_size=tok_right.vocab_size)).eval()
outputs_left = model(inputs_left['input_ids'], inputs_left['attention_mask'])
outputs_right = model(inputs_right['input_ids'], inputs_right['attention_mask'])

torch.allclose(outputs_left[0], outputs_right[0], atol=1e-5) # False
(outputs_left[0] - outputs_right[1]).max() # 0.3989
```

I'll dig into their fix and see if it will be easy to port over.

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2280 @@
+        emb = torch.arange(max_positions, dtype=torch.float).unsqueeze(1) * emb.unsqueeze(0)
+        return torch.sin(emb), torch.cos(emb)
+
+    def rotary(self, x):
+        n, d = x.size()
+        x1, x2 = torch.chunk(x, 2, dim=-1)
+        if self.sine is None or n > self.sine.size(0):
+            self.sine, self.cosine = RotaryRelativePositionalBias.get_sinusoid_embeddings(n, d)
+            self.max_positions = n
+        self.sine = self.sine.to(self._float_tensor)
+        self.cosine = self.cosine.to(self._float_tensor)
+
+        sin = self.sine[:n]
+        cos = self.cosine[:n]
+        return torch.cat([x1 * cos - x2 * sin, x2 * cos + x1 * sin], dim=1)
```
no single letter variables 😉 
let's find suitable names

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2280 @@
+
+        prediction_scores = self.lm_head(sequence_output)
+
+        lm_loss = None
+        if labels is not None:
+            # we are doing next-token prediction; shift prediction scores and input ids by one
+            shifted_prediction_scores = prediction_scores[:, :-1, :].contiguous()
+            labels = labels[:, 1:].contiguous()
+            loss_fct = CrossEntropyLoss()
+            lm_loss = loss_fct(shifted_prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
```
Should this model be `MegaLMHeadModel`

mnaylor5 replied:
Thank you! 🙌

ArthurZucker reviewed (commented):
Wow ! A lot of work, and I think we are almost there! I left a few nits here and there again, but should be ready soon. 🚀

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+# Gated cross-attention
+# removed before_attn_fn argument and unused static_kv
+# otherwise left as-is with the exception of hidden states
+class MegaGatedCrossAttention(nn.Module):
```
No problem at all! Thanks for understanding 🤗

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2311 @@
+        use_cache=False,
+    ):
+        """
+        Mega's self-attention block, which combines multi-headed EMA with traditional self-attention
```
Awesome! Thank you!

mnaylor5 commented on src/transformers/models/mega/configuration_mega.py:
```diff
@@ -0,0 +1,241 @@
+
+class MegaConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`MegaModel`] or a [`TFMegaModel`]. It is used to
```
Whoops, good catch! Changing locally

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2280 @@
+
+
+# utility for causal LM masking in the format that Mega expects
+def generate_causal_mask(seq_len):
```
This is addressed in the upcoming commit - I also realized that my old method was not handling the device. Since I haven't been able to test locally with a GPU, that had not triggered a test failure.

mnaylor5 committed:
documentation fixes, renaming variables in dropout and rotary positions, used built-in causal masking, encoders->layers in MegaModel, moved comments into docstrings

mnaylor5 committed:
style and quality fixes after latest updates, before rotary position ids

ArthurZucker commented:
Sure! Also it's not that important to have left padding in this PR, can be added in another PR!

mnaylor5 committed:
causal mask in MegaBlock docstring + added missing device passing

mnaylor5 committed:
Merge remote-tracking branch 'upstream/main' into add-mega

mnaylor5 commented:
Thanks @ArthurZucker! After digging into it, I do think it will require a pretty significant refactor to support left-padding in this PR. If you're comfortable with it, I agree that it could make sense in a new PR. I just added an entry in the `MegaBlock` docstring for the new `causal_mask` coming from the pretrained model's method, and added a missing `device` for the token type IDs. 

Also pulled latest changes from `main` to hopefully prevent whatever was causing the tests for exotic models to fail. I'm really happy with how this is looking, so let me know if there's anything else needed to move forward with this PR! Appreciate your comments and guidance on everything so far! :rocket:

ArthurZucker commented on tests/models/mega/test_modeling_mega.py:
```diff
@@ -0,0 +1,651 @@
+        self, config, input_ids, token_type_ids, input_mask, sequence_labels, token_labels, choice_labels
+    ):
+        config.use_chunking = True
+        config.chunk_size = (
+            input_ids.size(1) * 2
+        )  # we want the chunk size to be < sequence length, and the sequence length to be a multiple of chunk size
```
```suggestion
        # we want the chunk size to be < sequence length, and the sequence length to be a multiple of chunk size
        config.chunk_size = (input_ids.size(1) * 2)
```

mnaylor5 replied:
Applying this in the local commit because I'm not sure whether the parentheses are forcing the automated style checks to expand across lines

ArthurZucker commented on tests/models/mega/test_modeling_mega.py:
```diff
@@ -0,0 +1,651 @@
+
+        self.parent.assertEqual(result[0].shape, (self.batch_size, self.seq_length, self.hidden_size))
+
+    def check_chunking_shorter_sequence(
```
Nice!

ArthurZucker reviewed (approved):
Well done! 🔥 
I left a few comments here ans there about naming conventions and dosctrings, but this is very detailed, and love that you took the time to adress all of my comments! Thanks for bearing with me 😉

ArthurZucker commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2295 @@
+
+        # (sequence_length X batch_size X shared_representation_size + intermediate_size)
+        # split into two different tensors: one for Q/K usage and the other for gating self-attention
+        query_key, attention_gate = torch.split(
+            F.silu(query_key_gates), [self.config.shared_representation_size, self.config.intermediate_size], dim=-1
+        )
```
Let's also split this in 2 line (activation then this)

mnaylor5 replied:
Done in upcoming commit

sgugger commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2295 @@
+        return embeddings
+
+
+class SimpleRelativePositionalBias(nn.Module):
```
Should be prefixed by Mega

mnaylor5 replied:
Done in upcoming commit

sgugger commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2295 @@
+        return tile
+
+
+class RotaryRelativePositionalBias(nn.Module):
```
Should be prefixed by Mega

mnaylor5 replied:
Done in upcoming commit

sgugger commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2295 @@
+# original Mega repo used a parent class to organize normalization functions; Hugging Face's
+# preferred method is to use a dictionary. since these normalization classes require different
+# inputs, we handle this with lambda functions to return the instantiated class
+NORM2FN = {
+    "layernorm": lambda embedding_dim, eps, affine: nn.LayerNorm(embedding_dim, eps, elementwise_affine=affine),
+    "scalenorm": lambda embedding_dim, eps, affine: ScaleNorm(dim=-1, eps=eps, affine=affine),
+    "rmsnorm": lambda embedding_dim, eps, affine: MegaRMSNorm(embedding_dim, eps=eps, affine=affine),
+    "batchnorm": lambda embedding_dim, eps, affine: nn.BatchNorm1d(embedding_dim, eps=eps, affine=affine),
+    "syncbatchnorm": lambda embedding_dim, eps, affine: nn.SyncBatchNorm(embedding_dim, eps=eps, affine=affine),
+}
```
Using lambda functions here will make `MegaSequenceNorm` and then the whole model unpicklable I fear. It's probably better to have five if/else in the MegaSequenceNorm module.

ArthurZucker replied:
My bad! I was the one pushing for this! Miscalculated the pickling

mnaylor5 replied:
Ah, that makes sense! I originally had it as an if/else in the `MegaSequenceNorm.__init__`, so I'll just go back to that design

sgugger commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2295 @@
+ALL_LAYERNORM_LAYERS.append(MegaSequenceNorm)
+
+
+class MultiDimensionDampedEMA(nn.Module):
```
Should be prefixed by Mega

sgugger commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2295 @@
+        return outputs
+
+
+# Normalization modules
+# copied from original Mega repo without modification except variable names
+class ScaleNorm(nn.Module):
```
Also needs to prefixed by Mega

mnaylor5 replied:
On it! Thanks guys

sgugger commented on tests/models/mega/test_modeling_mega.py:
```diff
@@ -0,0 +1,651 @@
+        MegaForTokenClassification,
+        MegaModel,
+    )
+    from transformers.models.mega.modeling_mega import (
+        MEGA_PRETRAINED_MODEL_ARCHIVE_LIST,
+    )
```
Fits in one line.

mnaylor5 replied:
Done in upcoming commit

sgugger reviewed (commented):
Thanks a lot for all your work adding this model! My main comment is that all building blocks in the modeling file should be prefixed by Mega to avoid any name conflicts with other models.

mnaylor5 committed:
Apply suggestions from code review

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

mnaylor5 commented on src/transformers/models/mega/modeling_mega.py:
```diff
@@ -0,0 +1,2295 @@
+        return output
+
+
+class MovingAverageGatedAttention(nn.Module):
```
Working on this now and will add in a local commit to make sure it's updated wherever it's referenced

mnaylor5 commented on README.md:
```diff
@@ -377,6 +377,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
 1. **[MaskFormer](https://huggingface.co/docs/transformers/model_doc/maskformer)** (from Meta and UIUC) released with the paper [Per-Pixel Classification is Not All You Need for Semantic Segmentation](https://arxiv.org/abs/2107.06278) by Bowen Cheng, Alexander G. Schwing, Alexander Kirillov.
 1. **[mBART](https://huggingface.co/docs/transformers/model_doc/mbart)** (from Facebook) released with the paper [Multilingual Denoising Pre-training for Neural Machine Translation](https://arxiv.org/abs/2001.08210) by Yinhan Liu, Jiatao Gu, Naman Goyal, Xian Li, Sergey Edunov, Marjan Ghazvininejad, Mike Lewis, Luke Zettlemoyer.
 1. **[mBART-50](https://huggingface.co/docs/transformers/model_doc/mbart)** (from Facebook) released with the paper [Multilingual Translation with Extensible Multilingual Pretraining and Finetuning](https://arxiv.org/abs/2008.00401) by Yuqing Tang, Chau Tran, Xian Li, Peng-Jen Chen, Naman Goyal, Vishrav Chaudhary, Jiatao Gu, Angela Fan.
+1. **[MEGA](https://huggingface.co/docs/transformers/model_doc/mega)** (from Facebook) released with the paper [Mega: Moving Average Equipped Gated Attention](https://arxiv.org/abs/2209.10655) by Xuezhe Ma, Chunting Zhou, Xiang Kong, Junxian He, Liangke Gui, Graham Neubig, Jonathan May, and Luke Zettlemoyer.
```
Thanks for the review, @sgugger 🙏 I'll apply your suggestions wherever I can and make the rest in my local commit with the rest of @ArthurZucker's feedback

mnaylor5 committed:
Update README.md

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

mnaylor5 committed:
added Mega prefixes where missing, reverted MegaSequenceNorm to if-else, other module renaming requested in PR

mnaylor5 committed:
style and quality fixes + readme updates pointing to main

mnaylor5 committed:
Merge remote-tracking branch 'upstream/main' into add-mega

mnaylor5 commented:
Thanks again @ArthurZucker and @sgugger! Appreciate the feedback, and it should all be addressed in the latest changes 🤗

sgugger reviewed (approved):
Congrats on adding this new model to Transformers!

ArthurZucker commented:
Great working with you @mnaylor5 ! Congrats again on the merge 🔥

NielsRogge commented:
Congrats @mnaylor5 ! Feel free to share on social media and we'll amplify your post

mnaylor5 commented:
Thanks so much @ArthurZucker and @NielsRogge! I learned a ton through this process, and it's so rewarding to see my code in a library I use so much :heart:

I posted something here on LinkedIn a couple days ago - I'll tag you guys in the comments as well! 
https://www.linkedin.com/posts/mitchnaylor_mega-activity-7045103140890660864-9VOU
